### PR TITLE
Add functionality to _ftp_loader.py

### PR DESCRIPTION
**Pull request by chelseabright96**
-Created base class BaseFtpLoader
-Created classes FtpLoaderEnsemble and FTPLoaderNCBI inheriting from BaseFtpLoader
-Added pre-commit config
-Ran pre-commit on all files

### CODE REVIEW

File: oligo_designer_toolsuite/IO/_database.py [hunk 1]
```diff
@@ -1,139 +1,214 @@
...
+    """_summary_"""
 
-        self.species = 'human'
-        self.genome_assembly = 'GRCh38'
-        self.annotation_release = 'current'
+    def __init__(self) -> None:
 
+        self.species = "human"
+        self.genome_assembly = "GRCh38"
+        self.annotation_release = "current"
```
Comment (IsraMekki0) [inline oligo_designer_toolsuite/IO/_database.py:18]: Put these as default params instead

File: oligo_designer_toolsuite/IO/_database.py [hunk 2]
```diff
@@ -1,139 +1,214 @@
...
-    def download_annotation(self, species, genome_assembly, annotation_source, annotation_release, dir_output):
+            raise ValueError("Database file does not exist!")
+
+    def download_annotation(
+        self,
+        species,
+        genome_assembly,
+        annotation_source,
+        annotation_release,
+        dir_output,
```
Comment (IsraMekki0) [inline oligo_designer_toolsuite/IO/_database.py:35]: Replace all these with FTPLoader class (FTP)

File: oligo_designer_toolsuite/IO/_database.py [hunk 3]
```diff
@@ -1,139 +1,214 @@
...
+            warnings.warn(
+                f"No genome assembly defined. Using default assembly {self.genome_assembly}!"
+            )
             genome_assembly = self.genome_assembly
 
         if annotation_release is None:
-            warnings.warn(f'No ncbi annotation release defined. Using default release {self.annotation_release}!')
+            warnings.warn(
+                f"No ncbi annotation release defined. Using default release {self.annotation_release}!"
+            )
```
Comment (IsraMekki0) [inline oligo_designer_toolsuite/IO/_database.py:50]: Move warnings where the FTPLoader is instantiated

File: oligo_designer_toolsuite/IO/_database.py [hunk 4]
```diff
@@ -1,139 +1,214 @@
...
+        if annotation_source == "NCBI":
             FTP = ftp_loader.FTPLoaderNCBI(species, genome_assembly, annotation_release)
-        elif annotation_source == 'ensemble'
-            FTP = ftp_loader.FtpLoaderEnsemble(species, genome_assembly, annotation_release)
-        
+        elif annotation_source == "ensemble":
+            FTP = ftp_loader.FtpLoaderEnsemble(
+                species, genome_assembly, annotation_release
+            )
+
```
Comment (IsraMekki0) [inline oligo_designer_toolsuite/IO/_database.py:59]: Remove these lines

File: oligo_designer_toolsuite/IO/_database.py [hunk 5]
```diff
@@ -1,139 +1,214 @@
...
-    '''
+    """
     def _get_gene(self, file_sequence, file_annotation, dir_output):
         pass
 
     def _get_gene_CDS(self, file_sequence, file_annotation, blockSize, dir_output):
         pass
-    '''
-
+    """
```
Comment (IsraMekki0) [inline oligo_designer_toolsuite/IO/_database.py:78]: For the following, it makes more sense to me to have these classes:

- NcbiDbCreator
- EnsembleDbCreator
- CustomDbCreator

Each would implement two functions:
- createReferenceDB
- createOligoDB(OligoClass)

And the oligo class would be outside of IO, and would contain Oligo related functions
I think we should discuss this @lisa-sousa and @francesco-campi :)

**Review by IsraMekki0 — COMMENTED** [review]
Here are my comments, please let me know if it doesn't make sense to you.
Also, sorry, I included some comments on _database.py although not related to this pool request

**Review by lisa-sousa — APPROVED** [review]
pre-commit changes approved

**Comment by IsraMekki0** [comment]
One more change that just came to my mind is to update `get_params() -> ftp_directory_gtf, ftp_directory_fasta, ftp_file_gtf, ftp_file_fasta` to `get_params(file_type) -> ftp_directory, ftp_file` where the file_type is either "gtf" or "fasta". Then, the `download_and_decompress` function would just become `download_and_decompress(self, ftp_link, file_type)` and no need for `download_gtf` and `download_fasta` anymore. I find it ugly to put strings (you'd also need to check if file_type is correct, and accept lower and upper cases), but it helps in refactoring. If you don't want to add these changes, it's okay. Let me know, so I can merge.
Thanks!

**Pull request merged**